### PR TITLE
[SPARK-48356][SQL] Support for FOR statement

### PR DESCRIPTION
**Pull request by dusantism-db**
### What changes were proposed in this pull request?
In this PR, support for FOR statement in SQL scripting is introduced. Examples:

```
FOR row AS SELECT * FROM t DO
   SELECT row.intCol;
 END FOR;
```

```
FOR SELECT * FROM t DO
   SELECT intCol;
 END FOR;
```

Implementation notes:
As local variables for SQL scripting are currently a work in progress, session variables are used to simulate them.
When FOR begins executing, session variables are declared for each column in the result set, and optionally for the for variable if it is present ("row" in the example above). 
On each iteration, these variables are overwritten with the values from the row currently being iterated.
The variables are dropped upon loop completion.

This means that if a session variable which matches the name of a column in the result set already exists, the for statement will drop that variable after completion. If that variable would be referenced after the for statement, the script would fail as the variable would not exist. This limitation is already present in the current iteration of SQL scripting, and will be fixed once local variables are introduced. Also, with local variables the implementation of for statement will be much simpler.

Grammar/parser changes:
`forStatement` grammar rule
`visitForStatement` rule visitor
`ForStatement` logical operator

### Why are the changes needed?
FOR statement is an part of SQL scripting control flow logic.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
New tests are introduced to all of the three scripting test suites: `SqlScriptingParserSuite`, `SqlScriptingExecutionNodeSuite` and `SqlScriptingInterpreterSuite`.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (dejankrak-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:354]: To avoid code duplication, can we rewrite this case and the similar above ones as:

case c: WhileStatementContext
           | RepeatStatementContext
           | LoopStatementContext
           | ForStatementContext
        if Option(c.beginLabel()).isDefined &&
          c.beginLabel().multipartIdentifier().getText.toLowerCase(Locale.ROOT).equals(label)
        => true

Comment (dusantism-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:354]: I agree we need this refactor, however is it's not trivial as these classes don't have a common superclass with `beginLabel() `method. 
The syntax for what you're suggesting would be
```
case c @ (WhileStatementContext
| RepeatStatementContext
| LoopStatementContext
| ForStatementContext)
if Option(c.beginLabel()).isDefined &&
c.beginLabel().multipartIdentifier().getText.toLowerCase(Locale.ROOT).equals(label)
=> true
```
however this doesn't compile, due to the above reason. To add the superclass with begin/end labels requires grammar changes, i think this is best done as a separate task.

Comment (dejankrak-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:354]: Sure, makes sense, probably best to consider adding a base class with common methods as part of a separate PR (not a priority), to avoid further code duplication.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (davidm-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:322]: nit: Keep in mind Danko's PR for label checks, because it will probably merge before this, so just add the checks here as well.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingInterpreter.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingInterpreter.scala:139]: can't we just append `dropVariableExec` to the `collection` within `bodyExec`?
current approach seems like unnecessary nesting, which we should avoid if possible.
it would also make it equivalent to what we are doing for regular compounds.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:764]: for my knowledge, can you explain what does the case with the `Map` means exactly, i.e. when will this happen?
also, how did we check that this is the complete list of the relevant cases?

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:707]: when the variable name is not provided (just the query) what is the desired behavior? did we check this with Serge? I would say that at the moment query is useless and we cannot access the results in any way, but would expect there is some way to access the results that should be supported (otherwise the whole grammar wouldn't make sense)?

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingInterpreter.scala
Comment (dusantism-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingInterpreter.scala:139]: Sure, will change. For some reason i thought we did it like this for compounds.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (dusantism-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:707]: Yes, we should check with Serge.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (miland-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:722]: Why do we need to create this every time? Can we fill `variablesMap` once and then reuse it?

Comment (dusantism-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:722]: We need to create it every time because the map is different for every row in the result set. You can see we call it on the `currRow`.

**Comment by davidm-db** [comment]
Can we rebase first to include already merged changes regarding the label checks, logical plans, etc? And I'll review afterwards again?

**Comment by dusantism-db** [comment]
@davidm-db @miland-db Rebased, you can review again

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:134]: this doesn't seem right, and definitely won't be right in the future (i.e. not all data frames will have/need to be executed immediately).
with execution v1 (that @miland-db) is working on, we will figure out the proper set of flags needed and where/when to update them.
for now, I would say to do something like:
```
execNode.buildDataFrame(session).collect()
execNode.isExecuted = true
```
instead of setting the `isExecuted` in the function here

File: sql/core/src/test/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNodeSuite.scala
Comment (davidm-db) [inline sql/core/src/test/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNodeSuite.scala:916]: we don't have drop var statements here due to the fact that they are dropped in `handleLeaveStatement`?
this is the thing we talked about that will be properly resolved once the proper execution and scopes are introduced?

File: sql/core/src/test/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNodeSuite.scala
Comment (davidm-db) [inline sql/core/src/test/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNodeSuite.scala:121]: maybe do a `statement: SingleStatementExec if statement.logicalPlan.isInstanceOf[DropVariable] => "DropVariable"` just to be more strict with the tests?

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:686]: food for thought: does DataFrame have a mechanism to partially collect the data so we don't collect all the results in memory? since we are already using the caching concept, this would be easy to add to the logic of `cachedQueryResult`.

quickly researching, we can do something like:
```sliced_df = df.offset(starting_index).limit(ending_index - starting_index)```
but there might be something better...

I wouldn't block the PR on this, but I think we definitely need to consider something like this for a follow-up.

cc: @cloud-fan @MaxGekk

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:701]: maybe have a `cacheSize` that's a class member and set in `cachedQueryResult()`?

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:794]: nit: add one empty line above, I started looking for where this case disappeared compared to the previous iterations 😂 it'll be a bit cleaner

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:353]:
```suggestion
      ctx: ForStatementContext,
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:354]:
```suggestion
      labelCtx: SqlScriptingLabelContext): ForStatement = {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:391]: you can express it as

```
if Option(c.beginLabel).map { b =>
  b.multipartIdentifier().getText...)
  }.getOrElse(false)
```

this way you don't have to repeat the name of the field.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SqlScriptingLogicalPlans.scala
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SqlScriptingLogicalPlans.scala:294]: you can pattern-match the two children:

```
newChildren match {
  case (s: SingleStatement, c: CompoundBody) =>
    ForStatement(s, variableName, c, label)
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNodeSuite.scala
Comment (dusantism-db) [reply sql/core/src/test/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNodeSuite.scala:916]: Yes, that's right. In this case the variables are dropped immediately when the leave statement is encountered, instead of the usual behavior which is to return the dropVariable exec nodes from the iterator.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (dusantism-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:701]: I don't think this is necessary, as the queryResult is a fixed size array, meaning we don't calculate the length here.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (dusantism-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:686]: That makes sense, currently the entire result is collected to the driver so it would be problematic if the result size is too large. We should definitely follow up on this

Comment (davidm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:685]: Let's see what Wenchen and Max have to say and maybe create a follow-up work item so we don't forget it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SqlScriptingLogicalPlans.scala
Comment (davidm-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SqlScriptingLogicalPlans.scala:293]:
```suggestion
  override protected def withNewChildrenInternal(
      newChildren: IndexedSeq[LogicalPlan]): LogicalPlan = newChildren match {
```

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:715]: should we set `isExecuted = true` here as well? or it is not important since we don't return it anywhere?
same question for the line 722 as well?

**Review by davidm-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (dusantism-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:715]: I don't think it's important since the execs are not persisted or used anywhere else.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:685]: There is a `df.toLocalIterator()`. Under the hood, it launches jobs for RDD partitions one by one, so at most only the data of one Partition will be collected to the Spark driver at the same time.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:706]: what will be returned by this iterator? The `query`? The internal create/set/drop session variables command?

Comment (dusantism-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:706]: The iterator doesn't return the query/create/set variables. It returns statements from the body, and after all iterations are complete, it returns the drop variables commands, one by one. When local variables are introduced, the iterator will only return statements from the body.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:691]: awesome, the approach looks good to me now!
it seems like we can simplify this further though, i.e. we don't need a function and helper vars now?

would something simple like this work?
```suggestion
  private lazy val queryResultIter: util.Iterator[Row] = {
    val queryResult = query.buildDataFrame(session).toLocalIterator()
    query.isExecuted = true
    queryResult
  }
```

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (davidm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala]: I think we can simplify this a bit further, commented it [here](https://github.com/apache/spark/pull/48794/files#r1854400972).
Otherwise, the idea is really cool. You already have my approval.

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (dusantism-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:691]: @davidm-db So i tried something like this already, but the problem is that lazy val can't be reset in `reset()`. With lazy val the query will only be executed once, and there is no way to make it execute again in the case the for statement needs to be executed again, for example if it's nested in another loop.

Comment (davidm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:691]: ooh, forgot completely about the `reset`... makes sense, thx!

File: sql/core/src/test/scala/org/apache/spark/sql/scripting/SqlScriptingInterpreterSuite.scala
Comment (dejankrak-db) [inline sql/core/src/test/scala/org/apache/spark/sql/scripting/SqlScriptingInterpreterSuite.scala:2541]: Can we also add a test case with a global SQL variable, and then iterate through the table using FOR loop and add a value for a specific column to the variable, with validating in the end that the variable value equals to the sum of all values for that column?

**Review by dejankrak-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:712]: @miland-db can you confirm whether we should return `query` or not here? Should we treat it the same as IF/WHILE condition and return it in the iterator?

File: sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala:700]: let's use `SparkPlan#executeToIterator()`, which returns `InternalRow`, so that we can save the cost of data conversion in `createExpressionFromValue`

**Comment by cloud-fan** [comment]
We can address comments later as they are kind of improvement. Let me merge it first. Thanks!

**Comment by dusantism-db** [comment]
@cloud-fan Great, thanks. I will create follow up tasks for the improvements you suggested.

**Pull request closed without merging**
